### ISSUE TITLE
Own handwriting (really need help)

### ISSUE THREAD

**Issue by Gerrit02** [makeyourownneuralnetwork/makeyourownneuralnetwork#46]
@makeyourownneuralnetwork or somebody else,
Please help. I really do not understand the difference... in way of working?
I did a small test with the own handwriting number 4 (the number you provide Rashid in your own handwriting numbers).

I did the test Rashid with your code part3_neural_network_mnist_and_own_data.ipynb (epoch = 10 and trainingset = mnist_train_100.csv

Here there is a match with the lable 4 (see attach)

Then I also did the test with your code (part2_neural_network_mnist_data.ipynb) where you provide the testset as dataset (eg: mnist_test_10.csv)
I put also the epoch = 10 and trained the network with mnist_train_100.csv
So in this case I converted the image.png into a csv file with 784 pixels and 4 in front as the correct label.
I called my testset mnist_test_03.csv

When I run the code I've got no match. (See attach)!

For once and forever can somebody do the same and debug why in one case it is working and in the other not. I'm really get disappointed in let the neural network work with data from own handwriting :-(

As you can see both labels are the same, and the network has been trained with same set of trainings data and epoch.
@makeyourownneuralnetwork can you please do the same test or can you explain. I realy like to know what is hapening. Tnx a lot.

![fournotworking](https://user-images.githubusercontent.com/34437030/37429305-2e5bc7be-27cf-11e8-989c-51ac01969704.JPG)
![fourworking](https://user-images.githubusercontent.com/34437030/37429306-2e8d4744-27cf-11e8-85c7-f3d128776d29.JPG)

**Comment by makeyourownneuralnetwork** [comment]
Let me look at this over the weekend. I'm a little busy this week.

**Comment by Gerrit02** [comment]
ok thank you!

**Comment by makeyourownneuralnetwork** [comment]
Hi @Gerrit02 

please can you explain how you turn the PNG image into CSV? that might be the cause as you may be inverting the greyscale values.

most software will have 255 as white but here we use 0 as white.

also the full training set rather than the shorter 100 is better .. the 100 is for just testing the code but not for actual classification.

**Comment by Gerrit02** [comment]
Hello @makeyourownneuralnetwork,
thank you for taking the time to look at my problem.
Two remarks.
1) Remember the training set you use in the code part3_neural_network_mnist_and_own_data.ipynb is also the 100 set. And also put the same epoch namely 10. So no difference what that concerns.

2) Secondly, the conversing, I did as you explained in your book. As you can see above the two labels are also identical. I think when you not inversed the greyscale the black and white pixels should also inversed in the labels no, but I can be wrong.

So, the values in my csv are exactly the same as the values you have for your singel number 4 label you use in the code part3. 

What I propose, please try it once, convert your label4.png and put the values in excel file and save as csv. Don't forget to put a 4 in front of all the pixel values. Then use this csv file in your part2 code just as you use the testset.csv or testset_10.csv file and run the code. It does no matter you train the network with the full training set or with the 100set. In both case's you have no match.

If it's working at your site, you may allways sent me the content of the csv file of your number 4 but I think nothing is wrong with that. 

Thanks in advance, I really like to know why it is recognizing your number 4 in the part 3 code but not in the part 2 code.

**Comment by Gerrit02** [comment]
@makeyourownneuralnetwork 
did you find some time in meanwhile to do the test (see above)?
Save your number 4 label as file.csv and run it with part2_neural_network_mnist_data.ipynb.

**Comment by makeyourownneuralnetwork** [comment]
see attached the CSV saved from the "4" image 

[4.csv.zip](https://github.com/makeyourownneuralnetwork/makeyourownneuralnetwork/files/1862299/4.csv.zip)

this is prepared from the image data by mapping the grey values (0 to 1) to (0 to 255)

see the screenshot below showing that when the test data is only this single-record 4.csv the neural network correctly classifies it as "4"
<img width="1137" alt="screen shot 2018-03-30 at 00 51 34" src="https://user-images.githubusercontent.com/17411198/38118761-dd32dbe2-33b4-11e8-9acb-b8a19f01c25a.png">

I hope this is helpful.

**Comment by Gerrit02** [comment]
@makeyourownneuralnetwork 
thx, I will try it out. But anyway I don't know where you come from the pixels 2.55, etc...
If you run your part3 code, the content of the pixels is shown as in my mnist_test_04Bis.zip
So where in your part3 code do you do the conversion, and is it possible to print the content of the pixels after the conversion then?

Another question, before conversion you have a value (for the first non white pixel) eg a value of 0.23905884.
In your 4.csv file your first non white pixel is 60,96. How do you come from 0.23905884 to 60,96?

Your sincerely,

**Comment by makeyourownalgorithmicart** [comment]
1. When loading images the greyscale values 0-255 are rescaled to be between 0 and 1. the code from part3 that does this is
```
# reshape from 28x28 to list of 784 values, invert values
img_data  = 255.0 - img_array.reshape(784)
    
# then scale data to range from 0.01 to 1.0
img_data = (img_data / 255.0 * 0.99) + 0.01
```
Yes, you can always print any value whenever you want, just use `print()`.


2. The data from the csv file is in the range 0 to 255. This needs to be converted to the range 0 to 1 as input to the network. 60.96/255 is 0.239 ...


3. The different between 0.01 and 2.55 is very small when the full range is 255. It's just how I prepared the image .. you could replace the 2.55 with 0.01 if you wanted to. What's important is where the numbers are in the range 0 - 255.

The question about black and white is answered in q(1). For this training set, the ink is 255 and empty is 0 (or close to these limits). If your image loads creates the opposite (such as 255 being white) then you need to invert these - but you need to check your image format and your image loading library as they can work differently.

**Comment by Gerrit02** [comment]
Hello @makeyourownalgorithmicart 

thank you for your answer. But still unclear for me :-(
First of all, in the part II code also the part to rescale the image data (greyscale inversion) is foreseen namely:
 inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01

Secondly what I do not understand is if you run the part 2 code and you take the mnist_test_10.csv there the values are put as eg: 0.01,0.01,...,0.23905884,0.01,0.01,... etc...
Then I take a csv with my own handwritten number (not my own but one of Rashid eg: nr 4) and there we must put the values as 2.55,2.55,...,60.96,...,2.55,... etc... 
Why? And as said the conversion code is also foreseen in part 2 ( inputs = (numpy.asfarray(all_values[1:]) / 255.0 * 0.99) + 0.01)!

Thirdly, also a point I do not understand, even if the csv contains the values 0.01,0.01,...,0.23905884,0.01,... or 2.55,2.55,...,60.96,2.55... etc... the label is always correctly printed. Meaning: my 4 consists of the black pixels and the background is white. But what I expect is when the greyscale is not inverted I expect to see the 4 in white pixels and the background in Black and that is not the case? Can you explain. (See labels 4 printed above, start of this topic)

**Comment by Gerrit02** [comment]
Hello @makeyourownalgorithmicart

I see in part II the greyscale is not inverted only the rescaling for the input.
But still my question remains why for the mnist_test_10.csv the values can be used as 0.01,0.01,...,0.23905884,0.01,0.01,... etc... 
And for the own image we have to use values as 2.55,2.55,...,60.96,...,2.55,... etc...

Also, why in both cases when I use as input the csv with the 0.01,0.01,...,0.23905884,0.01,0.01,... etc...  and as input the csv file with as values 2.55,2.55,...,60.96,...,2.55,... etc... my lable is printed in both cases the same. 4 with black pixels and background with white pixels. I expected at least an opposite image for both input csv files?

**Comment by Gerrit02** [comment]
@makeyourownneuralnetwork  @makeyourownalgorithmicart 
Sorry to boder again :-( nobody has the answer on above remark?
